### PR TITLE
COMPAT: restore shape for 'invalid' Index with nd array

### LINKED ISSUE #27775: IndexError: tuple index out of range after upgrade to 0.25
Root cause (in both cases using `df = pd.DataFrame({'a': [1, 2, 3]})`):

```
In [71]: pd.__version__  
Out[71]: '0.25.0'

In [73]: df.index[:, None]
Out[73]: Int64Index([0, 1, 2], dtype='int64')

In [74]: df.index[:, None].shape
Out[74]: (3,)
```

vs 

```
In [10]: pd.__version__  
Out[10]: '0.24.2'

In [13]: df.index[:, None] 
Out[13]: Int64Index([0, 1, 2], dtype='int64')

In [14]: df.index[:, None].shape
Out[14]: (3, 1)
```

So before, indexing with `[:, None]` (in numpy a way to add a dimension to get 2D array) actually resulting in Index with ndim of 2 (but which is of course inconsistent state of the Index object)

Matplotlib relied on this fact when an Index is passed to `plt.plot`, as reported in https://github.com/matplotlib/matplotlib/issues/14992

---

I have explained the issue [here ](https://stackoverflow.com/questions/57372557/matplotlib-drops-indexerror-tuple-index-out-of-range-after-upgrade)and [here ](https://github.com/matplotlib/matplotlib/issues/14992#issue-477267556)in details. Basically, after upgrading to the version `0.25` I got the error:

> IndexError: tuple index out of range

while attempting to plot a CSV file.

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
closes #27775

### CODE REVIEW

File: doc/source/whatsnew/v0.25.1.rst [hunk 1]
```diff
@@ -111,7 +111,7 @@ Plotting
 ^^^^^^^^
 
 - Added a pandas_plotting_backends entrypoint group for registering plot backends. See :ref:`extending.plotting-backends` for more (:issue:`26747`).
--
+- Fix compatibility issue with matplotlib when passing a pandas ``Index`` to a plot call (:issue:`27775`).
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v0.25.1.rst]: I like that this is vague about the underlying issue. I don't want people relying on Index.shape being n-d.

**Review by TomAugspurger — COMMENTED** [review]
@tacaswell will matplotlib be able to track down the places that need updating, to cast Index objects to an ndarray earlier? Would it help if we put an (invisible by default) DeprecationWarning in here?

**Comment by jorisvandenbossche** [comment]
>  Would it help if we put an (invisible by default) DeprecationWarning in here?

What would you deprecate? It's creating the 2D Index that we should deprecate, not necessarily accessing its shape. 
But deprecating that should go on master, not for 0.25.1 I think. 

To be clear, I only see this as a fix for 0.25.1, and we should further discuss how we want to fix this long term on master (convert to ndarray, or deprecate/raise error)

**Comment by TomAugspurger** [comment]
> What would you deprecate?

Not really a deprecation. Just a "hey this is wrong" warning that's invisible by default in `Index.shape`. I just don't know how matplotlib is going to catch this otherwise (though maybe they have a central place where they do their conversion to NumPy).

**Comment by jorisvandenbossche** [comment]
To be clear, do you propose that instead of this PR? (or as something that should be added to this PR?)

(for me that is again part of a separate longer term solution discussion)

**Comment by jreback** [comment]
seems ok to me, I think we should raise starting in 1.0.0 for this on consruction; its invalid.

**Comment by TomAugspurger** [comment]
>

**Comment by tacaswell** [comment]
@TomAugspurger Yes: https://github.com/matplotlib/matplotlib/pull/15007 although you may get a fix out before we do though :woman_shrugging: .

**Comment by tacaswell** [comment]
@jreback What type would you raise?  If it is `TypeError` or `IndexError` it will "just work" with Matplotilb.

**Comment by tacaswell** [comment]
Turns out this used to be extra goofy:

```python

In [13]: import pandas

In [14]: pandas.__version__
Out[14]: '0.24.2'

In [15]: import pandas as pd

In [16]: df = pd.DataFrame([1, 2])

In [17]: df.index[:, None].shape
Out[17]: (2, 1)

In [18]: df.index[:, None].ndim
Out[18]: 1

```

**Comment by jorisvandenbossche** [comment]
@tacaswell yes, it was buggy, it is still buggy, we just changed the way it was buggy (and this PR restores the previous way of buggy)

But let's keep all that discussion about additional fixes, returning array or erroring, .. for a proper issue about it (I will open one).

**Comment by jorisvandenbossche** [comment]
Opened an issue here: https://github.com/pandas-dev/pandas/issues/27837

**Pull request merged**